components/req_cpu/src/linux.rs

```diff
@@ -0,0 +1,690 @@
+        (!prev_ptr.is_null()).then(|| unsafe { Arc::from_raw(prev_ptr as _) })
+    }
+}
+
+struct LocalReqTag {
+    is_set: Cell<bool>,
+    shared_ptr: SharedReqTagPtr,
+}
+thread_local! {
+    static CURRENT_REQ: LocalReqTag = {
+        let thread_id = unsafe { libc::syscall(libc::SYS_gettid) as libc::pid_t };
+
+        let shared_ptr = SharedReqTagPtr::default();
+        THREAD_REGISTRATION_CHANNEL.0.send(ThreadRegistrationMsg {
+            thread_id,
+            shared_ptr: shared_ptr.clone(),
+        }).ok();
+
+        LocalReqTag {
+            is_set: Cell::new(false),
+            shared_ptr,
+        }
+    };
+}
+
+#[derive(Default)]
+pub struct Guard {
+    // A trick to impl !Send, !Sync
+    _p: PhantomData<*const ()>,
+}
+impl Drop for Guard {
+    fn drop(&mut self) {
+        CURRENT_REQ.with(|s| {
+            while s.shared_ptr.take().is_none() {}
+            s.is_set.set(false);
+        });
+    }
+}
+
+pub struct CollectorHandle {
+    id: CollectorId,
+}
+impl Drop for CollectorHandle {
+    fn drop(&mut self) {
+        COLLECTOR_REGISTRATION_CHANNEL
+            .0
+            .send(CollectorRegistrationMsg::Unregister { id: self.id })
+            .ok();
+    }
+}
+
+lazy_static! {
+    static ref PID: pid_t = unsafe { libc::getpid() };
+    static ref CLK_TCK: libc::c_long = unsafe { libc::sysconf(libc::_SC_CLK_TCK) };
+    static ref THREAD_REGISTRATION_CHANNEL: (
+        Sender<ThreadRegistrationMsg>,
+        Receiver<ThreadRegistrationMsg>
+    ) = unbounded();
+    static ref COLLECTOR_REGISTRATION_CHANNEL: (
+        Sender<CollectorRegistrationMsg>,
+        Receiver<CollectorRegistrationMsg>
+    ) = unbounded();
+}
+struct ThreadRegistrationMsg {
+    thread_id: pid_t,
+    shared_ptr: SharedReqTagPtr,
+}
+enum CollectorRegistrationMsg {
+    Register {
+        collector: Box<dyn Collector>,
+        id: CollectorId,
+    },
+    Unregister {
+        id: CollectorId,
+    },
+}
+
+struct ReqCpuRecorder {
+    config: RecorderConfig,
+
+    thread_stats: HashMap<pid_t, ThreadStat>,
+    current_window_records: RequestCpuRecords,
+
+    last_collect_instant: Instant,
+    last_gc_instant: Instant,
+
+    collectors: HashMap<CollectorId, Box<dyn Collector>>,
+}
+
+struct ThreadStat {
+    shared_ptr: SharedReqTagPtr,
+    prev_tag: Option<Arc<RequestTag>>,
+    prev_stat: pid::Stat,
+}
+
+impl ReqCpuRecorder {
+    pub fn new(config: RecorderConfig) -> Self {
+        let now = Instant::now();
+
+        Self {
+            config,
+
+            last_collect_instant: now,
+            last_gc_instant: now,
+
+            thread_stats: HashMap::default(),
+            current_window_records: RequestCpuRecords::default(),
+
+            collectors: HashMap::default(),
+        }
+    }
+
+    pub fn handle_collector_registration(&mut self) {
+        let mut should_reset = false;
+        loop {
+            while let Ok(msg) = COLLECTOR_REGISTRATION_CHANNEL.1.try_recv() {
```

Comment:
Do we need to add sleep in case of no collector is registered?

Reply:
In that case, the following  `COLLECTOR_REGISTRATION_CHANNEL.1.recv()` works like sleep.